water_neuron/src/icrc21.rs

```diff
@@ -0,0 +1,166 @@
+    pub utc_offset_minutes: Option<u16>,
+    pub language: String,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum DisplayMessageType {
+    GenericDisplay,
+    LineDisplay {
+        characters_per_line: u16,
+        lines_per_page: u16,
+    },
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ConsentInfo {
+    pub metadata: ConsentMessageMetadata,
+    pub consent_message: ConsentMessage,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum ConsentMessage {
+    LineDisplayMessage { pages: Vec<LineDisplayPage> },
+    GenericDisplayMessage(String),
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct LineDisplayPage {
+    lines: Vec<String>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum Icrc21Error {
+    GenericError {
+        description: String,
+        error_code: u64,
+    },
+    InsufficientPayment(ErrorInfo),
+    UnsupportedCanisterCall(ErrorInfo),
+    ConsentMessageUnavailable(ErrorInfo),
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct ErrorInfo {
+    pub description: String,
+}
+
+pub fn icrc10_supported_standards() -> Vec<StandardRecord> {
+    vec![
+        StandardRecord {
+            name: "ICRC-21".to_string(),
+            url: "https://github.com/dfinity/wg-identity-authentication/blob/main/topics/ICRC-21/icrc_21_consent_msg.md".to_string(),
+        }
+    ]
+}
+
+pub fn icrc21_canister_call_consent_message(
+    request: ConsentMessageRequest,
+) -> Result<ConsentInfo, Icrc21Error> {
+    if request.arg.len() > MAX_CONSENT_MESSAGE_ARG_SIZE_BYTES as usize {
+        return Err(Icrc21Error::UnsupportedCanisterCall(ErrorInfo {
+            description: format!(
+                "The argument size is too large. The maximum allowed size is {} bytes.",
+                MAX_CONSENT_MESSAGE_ARG_SIZE_BYTES
+            ),
+        }));
+    }
+
+    let metadata = ConsentMessageMetadata {
+        language: "en".to_string(),
+        utc_offset_minutes: request.user_preferences.metadata.utc_offset_minutes,
+    };
+
+    let message = match request.method.parse::<Icrc21Function>().map_err(|err| {
+        Icrc21Error::UnsupportedCanisterCall(ErrorInfo {
+            description: format!(
+                "The call provided is not supported: {}.\n Supported calls for ICRC-21 are {:?}. \n Error: {:?}", 
+                request.method,
+                Icrc21Function::iter().map(|f|f.to_string()).collect::<Vec<String>>(),
+                err
+            ),
+        })
+    })? {
+        Icrc21Function::Stake =>  {
+            let arg = Decode!(&request.arg, ConversionArg).map_err(|e| Icrc21Error::UnsupportedCanisterCall(ErrorInfo {
+                description: format!("Failed to decode ConversionArg: {}", e)
+            }))?;
+            let icp_amount = format!("{}", DisplayAmount(arg.amount_e8s));
+            match arg.maybe_subaccount {
+                Some(subaccount) => format!("Convert {icp_amount} ICP to nICP at the current exchange rate. Specified subaccount: {subaccount:?}."),
```

Comment:
```suggestion
            let icp_amount = format!("{}", DisplayAmount(arg.amount_e8s));
            match arg.maybe_subaccount {
                Some(subaccount) => format!("Convert {} ICP to nICP at the current exchange rate. Specified subaccount: {}.", DisplayAmount(icp_amount), hex::encode(subaccount))),
```

This comment applies for unstake as well.